src/witness/advancing_range.rs

```diff
@@ -0,0 +1,71 @@
+use std::ops::Range;
+
+/// Used for finding ranges in an array without having to iterate over the entire array every time.
```

Comment:
please improve this comment..

Explain a little bit more what you mean by 'ranges' - as suddenly there are some 'cycles' involved, right? and there is an assumptino about the contents of the data.

Reply:
I don't like it very much either and am open to suggestions. I will explain how I got here.

I initially assumed that every array is [(u32, T)] but that turned out to not be the case. Instead, cycle number they are sorted by is either the first in a tuple or the only value, as seen in the trait implementations.

The other logic is otherwise very generic but the specific way that the number is gotten out of the array makes it very specific to this use case.

Reply:
Added a bit more explanation and renamed it to TupleFirst.

Reply:
@EmilLuta I did yet another pass on the docs.

Reply:
looks a lot better !!

I'd also mention, that the optimisation comes from the fact, that each range that you're asking for has to be 'after' the range that you asked in the past.